tools/autograd/gen_python_functions.py

```diff
@@ -656,180 +653,152 @@ def method_def(name, declarations, is_python_method, module):
+    for i1, overload1 in enumerate(grouped_overloads):
+        for i2, overload2 in enumerate(grouped_overloads):
+            if is_smaller(overload1.signature, overload2.signature):
                 larger_than[i1].add(i2)
 
     if not larger_than:
-        return grouped_decls
-
-    # Use a topological sort to sort decls according to the partial order.
-    sorted_deps = [(i, decl) for i, decl in enumerate(grouped_decls)
-                   if i not in larger_than]
-    for i, decl in sorted_deps:
-        for i2 in sorted(larger_than.keys()):
-            larger = larger_than[i2]
+        return tuple(grouped_overloads)
+
+    # Use a topological sort to sort overloads according to the partial order.
+    sorted_ids: List[int] = \
+        list(filter(lambda x: x not in larger_than, range(len(grouped_overloads))))
+    for i in sorted_ids:
+        for j in sorted(larger_than.keys()):
+            larger = larger_than[j]
             larger.discard(i)
             if not larger:
-                del larger_than[i2]
-                sorted_deps.append((i2, grouped_decls[i2]))
-
-    return [decl for i, decl in sorted_deps]
-
-
-#
-# python signature codegen
-#
-
-def get_python_signature(declaration, is_python_method, skip_outputs=False):
-    return declaration['python_signature'].signature_str(skip_outputs=skip_outputs)
-
-
-#
-# op args to python parsed args transform
-#
-
-def get_python_argc(decl):
-    return len(decl['python_signature'].arguments())
-
-
-def get_python_output_index(decl):
-    ps: PythonSignature = decl['python_signature']
-    return len(ps.input_args) + len(ps.input_kwargs)
-
-
-#
-# declaration derived props, utils, etc.
-# declarations are dicts loaded from Declarations.yaml,
-# passed to our codegen methods by callers in gen_autograd
-#
-
-def is_output(arg):
-    return arg.get('output', False)
-
-
-def has_outputs(declaration):
-    return any([is_output(arg) for arg in declaration['arguments']])
-
-
-def is_torch_function(declaration):
-    return 'namespace' in declaration['method_of']
-
-
-def is_nn_module_function(declaration):
-    return declaration.get('python_module') == 'nn'
-
-
-def is_fft_module_function(declaration):
-    return declaration.get('python_module') == 'fft'
-
-
-def is_linalg_module_function(declaration):
-    return declaration.get('python_module') == 'linalg'
-
-
-def op_name(declaration):
-    name = declaration['name']
-    if has_outputs(declaration):
-        if not name.endswith("_out"):
-            raise RuntimeError(
-                '{} has output params, expecting name ending with \'_out\''.
-                format(declaration['name']))
-        return name[:-4]
-    else:
-        if name.endswith("_out"):
-            raise RuntimeError(
-                '{}: name ends with \'_out\', expecting output params'.
-                format(declaration['name']))
-        return name
+                del larger_than[j]
+                sorted_ids.append(j)
```

Comment:
Also the old code's fault, but mutating lists as I iterate over them always gives me the jeebies

Reply:
> Also the old code's fault, but mutating lists as I iterate over them always gives me the jeebies

ok - I've simply changed it to iterating through index indirectly instead of sorted_ids directly - not sure it hurts readability :-)